polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -789,362 +665,335 @@ impl FragmentTree {
+		relay_parent: &Hash,
+		parent_head_hash: Hash,
+		output_head_hash: Option<Hash>,
+	) -> PotentialAddition {
+		// If we've got enough candidates for the configured depth, no point in adding more.
+		if self.chain.len() > self.scope.max_depth {
+			return PotentialAddition::None
+		}
+
+		if !self.check_potential(relay_parent, parent_head_hash, output_head_hash) {
+			return PotentialAddition::None
+		}
+
+		let present_in_storage = storage.contains(candidate_hash);
+
+		let unconnected = self
+			.find_unconnected_potential_candidates(
+				storage,
+				present_in_storage.then_some(candidate_hash),
+			)
+			.len();
+
+		if (self.chain.len() + unconnected) < self.scope.max_depth {
+			PotentialAddition::Anyhow
+		} else if (self.chain.len() + unconnected) == self.scope.max_depth {
+			// If we've only one slot left to fill, it must be filled with a connected candidate.
+			PotentialAddition::IfConnected
+		} else {
+			PotentialAddition::None
+		}
+	}
+
+	// The candidates which are present in `CandidateStorage`, are not part of this chain but could
+	// become part of this chain in the future. Capped at the max depth minus the existing chain
+	// length.
+	// If `ignore_candidate`` is supplied and found in storage, it won't be counted.
+	pub(crate) fn find_unconnected_potential_candidates(
+		&self,
+		storage: &CandidateStorage,
+		ignore_candidate: Option<&CandidateHash>,
+	) -> Vec<CandidateHash> {
+		let mut candidates = vec![];
+		for candidate in storage.candidates() {
+			if let Some(ignore_candidate) = ignore_candidate {
+				if ignore_candidate == &candidate.candidate_hash {
+					continue
+				}
+			}
+			// We stop at max_depth + 1 with the search. There's no point in looping further.
+			if (self.chain.len() + candidates.len()) > self.scope.max_depth {
+				break
 			}
+			if !self.candidates.contains(&candidate.candidate_hash) {
+				if self.check_potential(
+					&candidate.relay_parent,
+					candidate.candidate.persisted_validation_data.parent_head.hash(),
+					Some(candidate.candidate.commitments.head_data.hash()),
+				) {
+					candidates.push(candidate.candidate_hash);
+				}
+			}
+		}
 
-			last_node = node;
+		candidates
+	}
 
-			next_node = match node {
-				NodePointer::Root => {
-					let children = self
-						.nodes
-						.iter()
-						.enumerate()
-						.take_while(|n| n.1.parent == NodePointer::Root)
-						.map(|(index, node)| (NodePointer::Storage(index), node.candidate_hash))
-						.collect::<Vec<_>>();
+	// Check if adding a candidate which transitions `parent_head_hash` to `output_head_hash` would
+	// introduce a fork or a cycle in the parachain.
+	// `output_head_hash` is optional because we sometimes make this check before retrieving the
+	// collation.
+	fn is_fork_or_cycle(&self, parent_head_hash: Hash, output_head_hash: Option<Hash>) -> bool {
+		if self.by_parent_head.contains_key(&parent_head_hash) {
+			// fork. our parent has another child already
+			return true;
```

Comment:
Might want to run a round of `fmt`
```suggestion
			return true
```

Reply:
removed these specific occurrences. however, this doesn't seem a rule that rustfmt forbids as configured in the CI, since it's passing anyway and my local formatter does not change this either